src/main/scala/is/hail/stats/DownsampleCombiner.scala

```diff
@@ -0,0 +1,110 @@
+package is.hail.stats
+
+import org.apache.spark.sql.Row
+import is.hail.utils._
+
+class DownsampleCombiner(nDivisions: Int) extends Serializable {
+  type BinType = (Int, Int)
+  type PointType = (Double, Double, Seq[String])
+
+  var binToPoint: Map[BinType, PointType] = _
+  var initPoint: PointType = _
+
+  var left: Double = 0.0
+  var right: Double = 0.0
+  var bottom: Double = 0.0
+  var top: Double = 0.0
+
+  def toBinCoords(x: Double, y: Double, left: Double, right: Double, bottom: Double, top: Double, nDivisions: Int): (Int, Int) = {
+    val i = (((x - left) / (right - left)) * nDivisions).toInt
+    val j = (((y - bottom) / (top - bottom)) * nDivisions).toInt
+    (i, j)
+  }
+
+  def baseMerge(x: Double, y: Double, label: Seq[String]): DownsampleCombiner = {
+    assert(binToPoint != null)
+    binToPoint += (toBinCoords(x, y, left, right, bottom, top, nDivisions) -> (x, y, label))
+
+    this
+  }
+
+  def merge(x: Double, y: Double, label: Seq[String]): DownsampleCombiner = {
+    if (binToPoint != null) {
+      val bins = toBinCoords(x, y, left, right, bottom, top, nDivisions)
+      val xFactor = bins._1 / nDivisions
+      val yFactor = bins._2 / nDivisions
+      var collapse = false
+
+      if (math.abs(xFactor) > 1) {
+        collapse = true
+        if (xFactor < 0)
+          left = x
+        else
+          right = x
+      }
+
+      if (math.abs(yFactor) > 1) {
+        collapse = true
+        if (yFactor < 0)
+          bottom = y
+        else
+          top = y
+      }
+
+      if (collapse) {
+        val oldBinPoint = Map.empty ++ binToPoint
```

Comment:
No need for the `Map.empty ++`.  That makes a copy, but you can just use the old binToPoint since you're creating a new one from scratch.